migrations/lib/importer/discourse_db.rb

```diff
@@ -23,14 +29,28 @@ def copy_data(table_name, column_names, rows)
         @connection.transaction do
           @connection.copy_data(sql, @encoder) do
             sliced_rows.each do |row|
-              data = column_names.map { |c| row[c] }
+              if row[SKIP_ROW_MARKER]
+                skipped_rows << row
+                next
+              end
+
+              i = 0
+              while i < column_count
+                data[i] = row[column_names[i]]
+                i += 1
+              end
+
               @connection.put_copy_data(data)
+              inserted_rows << row
             end
           end
 
           # give the caller a chance to do some work when a batch has been committed,
           # for example, to store ID mappings
-          yield sliced_rows
+          yield inserted_rows, skipped_rows
+
+          inserted_rows.clear
+          skipped_rows.clear
```

Comment:
Micro optimization, but calling clear is slightly faster than creating new arrays in the loop